### PR TITLE
8338906: Avoid passing EnumDescs and extra classes to type switch methods that don't use them

### PR DESCRIPTION
**Pull request by cl4es**
This PR refactors SwitchBootstraps so that extra EnumDescs and classes are only passed into bootstraps when needed. Benchmarking shows that in many cases these are not needed, and avoiding passing them (via binding in the lists via `MethodHandle::insertArguments`) avoids some auxiliary MH combinator generation during bootstrap. 

Additional cleanups and refactoring further reduce bootstrap overhead and number of classes loaded.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8338906](https://bugs.openjdk.org/browse/JDK-8338906): Avoid passing EnumDescs and extra classes to type switch methods that don't use them (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20693/head:pull/20693` \
`$ git checkout pull/20693`

Update a local copy of the PR: \
`$ git checkout pull/20693` \
`$ git pull https://git.openjdk.org/jdk.git pull/20693/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20693`

View PR using the GUI difftool: \
`$ git pr show -t 20693`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20693.diff">https://git.openjdk.org/jdk/pull/20693.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20693#issuecomment-2307007478)

### CODE REVIEW

**Review by liach — APPROVED** [review]
Bytecode generation changes look good. Will wait for another compiler engineer to review.

/reviewers 2 reviewer

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java [hunk 1]
```diff
@@ -81,19 +82,27 @@ private SwitchBootstraps() {}
...
+    private static final MethodTypeDesc CHECK_INDEX_DESCRIPTOR =
+            MethodTypeDescImpl.ofValidated(ConstantDescs.CD_int, ConstantDescs.CD_int, ConstantDescs.CD_int);
+    private static final MethodTypeDesc MTD_TYPE_SWITCH = MethodTypeDesc.of(ConstantDescs.CD_int,
+            ConstantDescs.CD_Object,
+            ConstantDescs.CD_int);
+    private static final MethodTypeDesc MTD_TYPE_SWITCH_EXTRA = MethodTypeDesc.of(ConstantDescs.CD_int,
+            ConstantDescs.CD_Object,
+            ConstantDescs.CD_int,
+            CD_BiPredicate,
+            ConstantDescs.CD_List);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:97]: These can use `MethodTypeDescImpl​::ofValidated(…)`:
```suggestion
    private static final MethodTypeDesc MTD_TYPE_SWITCH = MethodTypeDescImpl.ofValidated(ConstantDescs.CD_int,
            ConstantDescs.CD_Object,
            ConstantDescs.CD_int);
    private static final MethodTypeDesc MTD_TYPE_SWITCH_EXTRA = MethodTypeDescImpl.ofValidated(ConstantDescs.CD_int,
            ConstantDescs.CD_Object,
            ConstantDescs.CD_int,
            CD_BiPredicate,
            ConstantDescs.CD_List);
```

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java [hunk 2]
```diff
@@ -81,19 +82,27 @@ private SwitchBootstraps() {}
     private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
     private static final boolean previewEnabled = PreviewFeatures.isEnabled();
 
+    private static final ClassDesc CD_BiPredicate = referenceClassDesc(BiPredicate.class);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:85]: This can use `ReferenceClassDescImpl​::ofValidated​(String)`, which also avoids going through `Class​::descriptorString()` for a statically known class descriptor:
```suggestion
    private static final ClassDesc CD_BiPredicate = ReferenceClassDescImpl.ofValidated("Ljava/util/function/BiPredicate;");
```

**Review by liach — APPROVED** [review]

**Comment by cl4es** [comment]
Thanks @liach 

Perhaps @lahodaj or @vicente-romero-oracle could take a look at this?

**Review by lahodaj — APPROVED** [review]
Looks reasonable to me.

**Comment by cl4es** [comment]
Thanks for reviewing!

/integrate

**Pull request closed without merging**